### PR TITLE
Add to_json Spark Function

### PR DESCRIPTION
**Pull request by duanmeng**
[to_json](https://spark.apache.org/docs/latest/api/sql/index.html#to_json)

to_json(expr[, options]) - Returns a JSON string with a given struct value

### CODE REVIEW

**Comment by jinchengchenghh** [comment]
@PHILO-HE Can you help take a look? I remember there are many options in Spark, which option should be supported at least?

File: velox/functions/sparksql/ToJson.cpp [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+};
+
+} // namespace
+
+VELOX_DECLARE_VECTOR_FUNCTION(
+    udf_to_json,
+    ToJsonFunction::signatures(),
+    std::make_unique<ToJsonFunction>());
+
+} // namespace facebook::velox::functions::sparksql
```
Comment (jinchengchenghh) [inline velox/functions/sparksql/ToJson.cpp:58]: Empty line in end

File: velox/functions/sparksql/CMakeLists.txt [hunk 2]
```diff
@@ -27,7 +27,8 @@ add_library(
   RegisterCompare.cpp
   Size.cpp
   SplitFunctions.cpp
-  String.cpp)
+  String.cpp
+  ToJson.cpp)
```
Comment (jinchengchenghh) [inline velox/functions/sparksql/CMakeLists.txt:31]: I would suggest to use name `json.cpp` to support following json functions in https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala

File: velox/functions/sparksql/ToJson.cpp (hunk 1 above)
Comment (duanmeng) [reply velox/functions/sparksql/ToJson.cpp]: Fixed

**Comment by duanmeng** [comment]
> @PHILO-HE Can you help take a look? I remember there are many options in Spark, which option should be supported at least?

@jinchengchenghh Thanks for your timely review. Those options are coupled with Spark's JVM-based `Jackson` lib. IMO we could reuse the `JsonCastOperator` of Velox's `JsonType` at least for now.

File: velox/functions/sparksql/CMakeLists.txt (hunk 2 above)
Comment (duanmeng) [reply velox/functions/sparksql/CMakeLists.txt]: Renamed.

File: velox/functions/sparksql/Json.cpp [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    JsonCastOperator::get()->castTo(
+        *args.at(0), context, rows, outputType, result);
+  }
+
+  static std::vector<std::shared_ptr<exec::FunctionSignature>> signatures() {
+    return {
+        exec::FunctionSignatureBuilder()
+            .typeVariable("T")
+            .returnType("varchar")
+            .argumentType("T")
```
Comment (PHILO-HE) [inline velox/functions/sparksql/Json.cpp:40]: To align with spark, it looks only array type & struct (map) type are acceptable. If some inner types like Date type, Timestamp type, etc. are not actually supported, we should not cover them in function signature.

**Comment by PHILO-HE** [comment]
> > @PHILO-HE Can you help take a look? I remember there are many options in Spark, which option should be supported at least?
> 
> @jinchengchenghh Thanks for your timely review. Those options are coupled with Spark's JVM-based `Jackson` lib. IMO we could reuse the `JsonCastOperator` of Velox's `JsonType` at least for now.

I think it's OK to just consider default options in the initial implementation. But I'm not sure whether this implementation is consistent with Spark's default behavior. For example, there is a Json option called `allowSingleQuotes` (true by default) in Spark to allow a field surrounded by single quotes. Maybe the inconsistent behaviors need to be well documented at least.

File: velox/functions/sparksql/Json.cpp (hunk 3 above)
Comment (duanmeng) [reply velox/functions/sparksql/Json.cpp:40]: Yes, you're right, only array, struct(row), and map type are acceptable in Spark. I used to try to add this constriction in the function signature. But I didn't find a way to add a signature for row type. It seemed that we must specify each column's type for row type. I tried `.argumentType("row")`, `argumentType("row(T)")` and `argumentType("row(T...)"` but  none work.

**Comment by duanmeng** [comment]
> > > @PHILO-HE Can you help take a look? I remember there are many options in Spark, which option should be supported at least?
> > 
> > 
> > @jinchengchenghh Thanks for your timely review. Those options are coupled with Spark's JVM-based `Jackson` lib. IMO we could reuse the `JsonCastOperator` of Velox's `JsonType` at least for now.
> 
> I think it's OK to just consider default options in the initial implementation. But I'm not sure whether this implementation is consistent with Spark's default behavior. For example, there is a Json option called `allowSingleQuotes` (true by default) in Spark to allow a field surrounded by single quotes. Maybe the inconsistent behaviors need to be well documented at least.

@PHILO-HE  Thanks for your help in reviewing. I will document it in the function's description file and comment on it in the source codes.

File: velox/functions/sparksql/Json.cpp (hunk 3 above)
Comment (duanmeng) [reply velox/functions/sparksql/Json.cpp:40]: @PHILO-HE PH FYI, https://github.com/facebookincubator/velox/discussions/6403

Comment (PHILO-HE) [reply velox/functions/sparksql/Json.cpp:40]: OK, let's keep your current impl. Thanks for clarification!

**Comment by brightpoker** [comment]
@duanmeng I tested this commit in gluten. I tried struct type but the result is different from spark. Here is my example, there is a field `struct_col`(struct<a:int, b:string>) in table t1. I insert '{"a":1,"b":"2"}' into t1 first, and then test 'select to_json(struct_col) from t1'. The result of velox is [1,"2"], but the result of spark is {"a":1,"b":"2"}.

**Pull request closed without merging**